### PR TITLE
Add support for deb822 APT sources

### PR DESCRIPTION
**Pull request by jps-help**
## Summary
Adds a new defined type, `apt::source_deb822`, and corresponding EPP template for managing [deb822 APT sources](https://manpages.debian.org/unstable/apt/sources.list.5.en.html#DEB822-STYLE_FORMAT).

Works alongside the existing `apt::source` type.

### TO DO
- Currently only supports specifying 'Signed-By' as a space-separated list of absolute file paths, but the deb822 spec also allows for writing the GPG ASCII armored key in-line with the .sources file.
- Unlike the existing `apt::source` type, doesn't currently support generating associated `apt::keyring` resources. Need to be created separately. 
- Proper spec tests.

## Additional Context
- deb822 APT sources have been supported by APT since v1.1
- Ubuntu 24.04 will ship with the Ubuntu repo configured in the newer deb822 style
- This module doesn't currently have a method for managing deb822 `.sources` files

## Related Issues (if any)
Mention any related issues or pull requests.

## Checklist
- [ ] 🟢 Spec tests.
- [ ] 🟢 Acceptance tests.
- [x] Manually verified. (For example `puppet apply`)

### CODE REVIEW

**Review by kenyon — APPROVED** [review]

**Comment by jps-help** [comment]
Inline ASCII GPG keys are still not working, but not necessary for an initial implementation.

Same with generating apt::keyring resources from within the apt::source_deb822 resource. But I don't really think this is necessary at all since you can just create the keyring resource separately.

Spec tests are done.

I'm happy with the state of this PR. Happy to hear people's thoughts/suggestions.

**Review by kenyon — APPROVED** [review]
Another area of improvement would be to use this in acceptance testing, otherwise 👍

**Comment by bastelfreak** [comment]
@jps-help thanks for the PR! Can you please rebase against main to get rid of the merge commit?

**Comment by jamesps-ebi** [comment]
> @jps-help thanks for the PR! Can you please rebase against main to get rid of the merge commit?

The fork has been rebased. Should be good to go.

**Review by bastelfreak — APPROVED** [review]

**Comment by jamesps-ebi** [comment]
Woops... I have accidentally reset my fork force pushed while testing another PR.

Luckily I still have my original deb822 branch. I'll fix and push again.

**Review by ptrsny — APPROVED** [review]
What exactly is your reason to add an additional resource instead of extending the apt::source resource with an Enum parameter with the values "list" and "source" or "legacy" and "deb822"?

Otherwise thanks for your work!

**Comment by jps-help** [comment]
> What exactly is your reason to add an additional resource instead of extending the apt::source resource

I think the main reason is just because of how different the two styles are. If you look at the current apt::source, many of the parameters take a single value, whereas, with deb822 parameters like `suites` and `components` can accept an array of values.
Sure we could just have one .sources file per suite or component of a repo, like with the current method, but I think it defeats the purpose of the newer deb822 style.

There might be a way we can extend the current type, I'm open to suggestions. My main focus was simply to add the new functionality without breaking the existing type at all.

**Comment by rrotter** [comment]
I'd suggest keeping this under `apt::source` for several reasons:

- `apt::source_deb822` would be a good name if this was an experimental interface, but deb822 is not "another way" it's the "new way", and (I think) will eventually be the *only way*. It should be presented as the default rather than an alternative, and especially not an alternative with a longer snake_cased_name.
- Having two resource types makes deleting sources worse. Imagine needing code like:
```
apt::source_deb822 { 'testing123': ensure => 'absent' }
apt::source { 'testing123': ensure => 'absent' }
```
- complicates migration: ideally upgrading to deb822 would be a one-line change to an existing `apt::source`; like adding `'type' => 'deb822'` or better still a zero-line change like omitting `'type' => 'legacy'`*
*this latter suggesting would probably need to wait for a major release since it's a breaking change)

> There might be a way we can extend the current type, I'm open to suggestions.

Indeed, I'd suggest adding an `Enum['legacy','deb822']` that probably defaults to 'legacy' for now. I like the idea of using the modern names for the fields (suites vs release, etc), but we could have both, and have one be an alias for the other. Could accept either an array or a string in the new name and continue accepting only a string in the old field name. That way there's the option of taking advantage of everything new that deb822 offers, but also the option of just switching over with at little change as possible.

**Comment by jps-help** [comment]
@rrotter All very good points. Let me take this back to the drawing board and try to implement with all these suggestions.

I think we will need a lot of new logic to handle the switch between the two versions, but in the end hopefully will be a better experience!

**Comment by jps-help** [comment]
Okay... I have merged some new changes that integrate the deb822 functionality into the existing `apt::source` type.

This is very much a work-in-progress and has several points to consider:
- I made the decision to add new parameter names for deb822 resources so they match with the content of the actual .sources file. This allows us to use different data types for those where appropriate.
- We might be able to allow an easy transition if we can find a way to have the new parameters use the old parameters as default values. I played with this but didn't implement because I found a few issues with that.
- Maintainers need to decide on the appropriate way to deprecate the existing method. Currently this change leaves the default as the current method. I tested a resource created using the current forge version, then upgrading to this version and saw no changes to the resource.
- I added a switch `source_format` which controls this behaviour. The default is 'legacy', and the new format is 'deb822' but maybe this should be 'list' and 'sources'? Or are we assuming there won't be any major change to apt sources for a long time?

Comments are greatly appreciated.

TLDR; I used a case statement to separate the functionality and added a new Enum parameter to control it.

**Comment by smoeding** [comment]
> * I added a switch `source_format` which controls this behaviour. The default is 'legacy', and the new format is 'deb822' but maybe this should be 'list' and 'sources'? Or are we assuming there won't be any major change to apt sources for a long time?

IMHO the 'list' and 'sources' options would be better since these are the file extensions that are eventually used. This makes it easier to remember the possible values.

In case that there will be another file format change in the future, the name of that format can easily be added here as allowed value. Otherwise 'deb822' would be the legacy format then.

**Comment by jps-help** [comment]
Finally had a chance to update this. I've used "list" and "sources" as the possible values for `source_format` and also re-used that parameter as the file_suffix later in the code.

I've also removed the separate `apt::source_deb822` type and some more failure conditions for deb822 style sources.

I think there will be more logic to add but can't think of anything right now. The only thing I see left are spec tests and README documentation but won't do those until we're happy with functionality.

**Comment by ptrsny** [comment]
Thank you for implementing the approach of only having one source resource.

As far as I can see the only thing I am missing in your implementation is @rrotter s [suggestion](https://github.com/puppetlabs/puppetlabs-apt/pull/1167#issuecomment-2191067763) of making the migration a one line change. This means, that the values `components <=> repos`, `suites <=> release`, `repo_trusted <=> allow_unsigned` and `uris <=> location` are (in my understanding) nearly interchangeable. So in other words if the parameter `repos` is set, and `source_format` is set to `source` I expect that it first tries to use the `components` parameter, but as long as that is not set it falls back to use `repos` (and that for all parameter combinations listed above).

I understand the idea of having the parameter names mapped to the corresponding keys of the format that is used. The concern I have with this is that the definition will explode every time a new format is added (hopefully not). Saying this I suggest that the doubled parameters should stay the old ones OR the new ones, but not both and "only" adapt the parameter types and improve the documentation accordingly.

Thanks again for your work on this. (And sorry for the late comment)

**Comment by jps-help** [comment]
> making the migration a one line change.

Sounds like a good idea. Thinking on this again, it doesn't make sense to add what are essentially duplicate parameters.

I think it will require a lot more logic though since we need to allow the original `String` value and `Array` value where appropriate. But it should all be possible.

**Comment by jps-help** [comment]
Migration between .list and .sources should now be much easier. 

I have consolidated the parameters used for both formats so they now mostly share the same parameters.
There are a few of exceptions:

1. $enabled - This is used exclusively by deb822 sources.
2. $types vs $include - Both are used to specify which types to include ('deb' and/or 'deb-src'). The newer $types param has a default value and I found that $include being a hash was needlessly complicated for what is essentially a Boolean choice.
3. $pin and $key - I've tried to implement these for deb822 sources but found it is quite complex. This can be revisited later but currently I just raise a warning that these are ignored for deb822 sources. We can always add apt::pin and apt::key resources separately anyway.

**Comment by jps-help** [comment]
Okay. All suggestions have been implemented where possible.

I have now added spec tests, updated README and rebased against MAIN since we were quite a bit out-of-sync.

Unless there are any more notes, I think this is ready.

File: manifests/params.pp
Comment (jps-help) [inline manifests/params.pp:1]: Looks like params.pp is being recreated by this PR. I'll fix it now.

**Comment by jps-help** [comment]
Is anything else needed to get this integrated? I'm happy to make changes as required, but I think this can be considered MVP right now.

**Comment by grant-veepshosting** [comment]
@bastelfreak is there any help I can contribute to this PR to expedite it (and #1169) to be merged and released?
These changes are critical to facilitate widespread adoption of key OS Ubuntu 24.04 including for a number of upstream modules.

**Comment by tmanninger** [comment]
@bastelfreak What's the state? Can i help to integrate?

**Comment by bastelfreak** [comment]
@grant-veepshosting @tmanninger this needs to be rebased before we can review and merge it. Is that something you could do?

**Comment by tmanninger** [comment]
@bastelfreak i have no write permissions to the source repo of this PR.
@jps-help can you rebase this PR? 
Should i create a new PR?

**Comment by rrotter** [comment]
@tmanninger I'm also very eager to see this moved along and would be happy to help in any way that I can.

**Comment by tmanninger** [comment]
When i am getting no feedback from @jps-help until tomorrow, i will fork this repo and create a new PR.
I need this PR as soon as possible.

**Comment by bastelfreak** [comment]
@tmanninger everybody can checkout the existing branch, rebase it an push it to their own fork for a new PR. You don't need write permissions in the original branch.

**Comment by jamesps-ebi** [comment]
@tmanninger @bastelfreak 
I'll look at rebasing this today :+1:

**Comment by jamesps-ebi** [comment]
> Can we add a default value in the init.pp for the format or use the new format when debian version >= 12 or ubuntu version >= 24 ?

I think let's try and get this out of the door first before we change any default behaviour. As it stands now, the changes shouldn't affect anyone's configuration unless they explicitly use the deb822 format.

**Comment by jamesps-ebi** [comment]
@bastelfreak @tmanninger 

I have rebased this PR. There was a horde of merge conflicts to deal with so I've done my best to get everything in-sync with upstream. Someone might want to run a diff to make sure nothing has regressed.

All spec tests are passing when run locally, but I'm going to try a couple of manual tests too.

**Comment by jps-help** [comment]
Looks good to me after manual testing. Feel free to do whatever other tests you want.

Let me know if anything needs changing.

File: manifests/setting.pp
Comment (bastelfreak) [inline manifests/setting.pp:52]:
```suggestion
  if $setting_type in ['list', 'pref', 'sources'] {
```

Comment (bastelfreak) [reply manifests/setting.pp:52]: this should be the best solution:
```puppet
$_priority = if $setting_type in ['list', 'pref', 'sources'] {
  ''
} else {
  $priority
}

**Comment by jps-help** [comment]
I saw main had some new commits, so I have rebased again.

**Comment by amitkarsale** [comment]
is this PR good to be merged?

**Review by david22swan — APPROVED** [review]
Changes look good to me, though as a previous comment stated, due to the size of the changes made an acceptance test to cover them would be preferrec

File: manifests/source.pp
Comment (smortex) [inline manifests/source.pp:245]: This feel wrong as `allow_unsigned => true` leads to `trusted => 'yes'`,

I am understanding this wrongly, or did you mean:
```suggestion
              'trusted'           => $allow_unsigned ? { false => 'yes', default => undef },
```

Comment (jps-help) [reply manifests/source.pp:245]: This is largely the same as what's in main currently: https://github.com/puppetlabs/puppetlabs-apt/blob/b565ce0b6d203eabb122d87e62db4e5457a6f4bc/manifests/source.pp#L219

I think it's correct as-is. 
`trusted => true` essentially results in "I trust this repo, do not fail even if the repo is unsigned".
and `trusted => false` results in no addition within the generated sources.list file, which by default means the source is not "trusted".

Comment (smortex) [reply manifests/source.pp:245]: Ah I see, thanks!  I updated #1212 which is based on your work to improve templates and tests accordingly.

**Review by malikparvez — APPROVED** [review]

**Comment by kenyon** [comment]
I think #1212 should have been merged instead of this one.

File: manifests/source.pp
Comment (ekohl) [inline manifests/source.pp:143]: This doesn't deal with `repos` being an array, does it?

Comment (jps-help) [reply manifests/source.pp:143]: Yes I think you might be right. I didn't really change anything to do with the original functionality. Basically I just dumped it into a case statement to separate it from the new  .sources format.

Since this has already been merged, it might be best to raise this as a new issue.

**Pull request merged**
